apps/todomvc/todomvc-web-components/dist/utils/script-loader.js

```diff
@@ -0,0 +1,62 @@
+import { requestIdleCallback } from "./request-idle-callback.js";
+
+function createScript({type}) {
+    return new Promise((resolve) => {
+        const scriptEl = document.createElement('script');
+        if (type !== "") scriptEl.type = type;
+        return resolve({ success: true, type: "createScript", scriptEl});
+    });
+}
+
+function addScript({scriptEl, location}) {
+    return new Promise((resolve) => {
+        if (location === 'head') {
+            document.head.appendChild(scriptEl);
+        } else {
+            document.body.appendChild(scriptEl);
+        }
+        return resolve({ success: true, type: "addScript"});
+    });
+}
+
+function initScript({scriptEl, url, strategy}) {
+    return new Promise((resolve, reject) => {
+      scriptEl.onload = () => resolve({success: true, type: "initScript"});
+      scriptEl.onerror = () => reject({success: false, type: "initScript"});
+
+      switch(strategy) {
+        case "lazyOnLoad":
+            requestIdleCallback(() => scriptEl.setAttribute('src', url));
+            break;
+        default:
+            scriptEl.setAttribute('src', url);
+      }
+    });
+}
+
+function buildScript({scriptEl, code}) {
+    return new Promise((resolve) => {
+        scriptEl.textContent = code;
+        return resolve({ success: true, type: "buildScript"});
+    });
+}
+
+export async function loadScript({ id, url = "", code = "", type = "", strategy = "default", location = "body", onError, onSuccess } = {}) {
+    // create a script element
+    const { scriptEl } = await createScript({type});
+
+    // add script to document
+    await addScript({scriptEl, location});
+
+    if (code !== "") {
+        // add inline code to script element
+        await buildScript({scriptEl, code});
+        onSuccess?.();
+    } else {
+        // load external url
+        const {success} = await initScript({scriptEl, url, strategy});
+        success ? onSuccess?.() : onError?.();
```

Comment:
I still find it a bit weird to mix the continuation-based approach (onXXX) with async-await here (but that might be just a matter of style).

I'd imagine something like this:
```
try {
  let scriptId = await loadScript(...);
  // onSuccess code goes here
} catch (error) {
  // onError code goes here
}
```

What't the advantage of passing back the `{success: true}` object?

Reply:
simplified and removed overhead that's not needed - lmk